src/utils/src/promise.js

```diff
@@ -0,0 +1,95 @@
+import Promise from 'es6-promise';
+
+/**
+ * @private
+ */
+const noop = function() {};
+
+/**
+ * @private
+ */
+function resolveWith(value) {
+  if (value && typeof value.then === 'function') {
+    return value;
+  }
+
+  return Promise.resolve(value);
+}
+
+/**
+ * @private
+ */
+export class Queue {
```

Comment:
@thomasconner can we use a good promise queue package instead of maintaining this ourselves?